### ISSUE TITLE
[BUG] Method matching positive on abstract classes which does not implements pointcuted method from interface

### ISSUE THREAD

**Issue by TheCelavi** [goaop/framework#335]
Consider usecase (precondition):

- Class/Interface structure: Interface I -> AbstractClass A implements I -> ConcreteClass C extends  A
- Interface defines method foo()
- Abstract class DOES NOT implements foo()
- ConcreteClass implements foo()

Consider that ALL stated Interface/Classes are within same namespace.

If matcher is execution(public *->foo()), abstract class gets weaved -> it is interpreted as if there is a method foo, but there is not.

Possible fix: check PHP Tokenizer and class metadata for better introspection of implemented methods.

**Comment by lisachenko** [comment]
Hi! I can see for the abstract class, that it doesn't contain an abstract method after proxy generation. because `ClassProxy`'s method `overrideMethod` has special temporary check at line 382:

```php
        // temporary disable override of final methods
        if (!$method->isFinal() && !$method->isAbstract()) {
            $this->override($method->name, $this->getJoinpointInvocationBody($method));
        }
```

This check prevents interception of final or abstract methods (including methods from interfaces).

But I can agree that pointcut currently matches abstract and final methods too. Not sure if it's good or bad, just possible to match abstract method. It can be used for example for introduction pointcuts to introduce new properties, methods or body for interfaces.

**Comment by TheCelavi** [comment]
> Not sure if it's good or bad, just possible to match abstract method.

I think it is bad, it should not match what can not be weaved. 

> It can be used for example for introduction pointcuts to introduce new properties, methods or body for interfaces.

Isn't that different story here? Introductions would introduce new Interfaces to a class, and via traits interfaces are implemented, right?

**Comment by lisachenko** [comment]
> Isn't that different story here? Introductions would introduce new Interfaces to a class, and via traits interfaces are implemented, right?

Yes, for example, TYPO3 just adds properties and methods into specific class. But I've decided to switch to traits and interfaces to have an ability to put breakpoints in traits. And it also gives me control over line numbering in the classes which was important.

 > I think it is bad, it should not match what can not be weaved.

Ok, let's fix this 👍

**Comment by lisachenko** [comment]
I haven't checked, but I think it will be possible to do this on grammar level by giving to the `ModifierMatcherFilter` negative filter for abstract modifier like that:

```php
$modifierMatcherFilter = new ModifierMatcherFilter();
$modifierMatcherFilter->notMatch(\ReflectionMethod::IS_ABSTRACT);
```

**Comment by lisachenko** [comment]
Fixed in #337
